mindsdb/interfaces/functions/to_markdown.py

```diff
@@ -0,0 +1,174 @@
+import base64
+from io import BytesIO
+import os
+from typing import Union
+from urllib.parse import urlparse
+
+import fitz  # PyMuPDF
+from markitdown import MarkItDown
+from openai import OpenAI
+import requests
+
+
+class ToMarkdown:
+    """
+    Extracts the content of documents of various formats in markdown format.
+    """
+    def __init__(self, use_llm: bool, llm_client: OpenAI = None, llm_model: str = None):
+        """
+        Initializes the ToMarkdown class.
+        """
+        # If use_llm is True, llm_client and llm_model must be provided.
+        if use_llm and (llm_client is None or llm_model is None):
+            raise ValueError('LLM client and model must be provided when use_llm is True.')
+        # Only OpenAI is supported for now.
+        if llm_client is not None and not isinstance(llm_client, OpenAI):
+            raise ValueError('Only OpenAI models are supported at the moment.')
+        # TODO: Add support for other LLMs?
+        self.llm_client = llm_client
+        self.llm_model = llm_model
+
+    def call(self, file_path_or_url: str) -> str:
+        """
+        Converts a file to markdown.
+        """
+        file_extension = self._get_file_extension(file_path_or_url)
+        file = self._get_file(file_path_or_url)
+
+        if file_extension == '.pdf':
+            return self._pdf_to_markdown(file)
+        elif file_extension in ['.jpg', '.jpeg', '.png', '.gif']:
+            return self._image_to_markdown(file)
+        else:
+            return self._other_to_markdown(file)
+
+    def _get_file(self, file_path_or_url: str) -> str:
+        """
+        Retrieves the content of a file.
+        """
+        parsed_url = urlparse(file_path_or_url)
+        if parsed_url.scheme in ('http', 'https'):
+            response = requests.get(file_path_or_url)
+            if response.status_code == 200:
+                return response
+            else:
+                raise RuntimeError(f'Unable to retrieve file from URL: {file_path_or_url}')
+        else:
+            with open(file_path_or_url, 'rb') as file:
+                return file.read()
+
+    def _get_file_extension(self, file_path_or_url: str) -> str:
+        """
+        Retrieves the file extension from a file path or URL.
+        """
+        parsed_url = urlparse(file_path_or_url)
+        if parsed_url.scheme in ('http', 'https'):
+            return os.path.splitext(parsed_url.path)[1]
```

Comment:
sometimes url doesn't content extention or file name for example https://arxiv.org/pdf/2412.20512
the type of file can be detected by content type headers